dlib/image_transforms/interpolation.h

```diff
@@ -229,94 +229,130 @@ namespace dlib
+                return true;
+            }
+        };
+
+        template<typename gcc_happy>
+        struct impl_t<3, gcc_happy>
         {
-            COMPILE_TIME_ASSERT(pixel_traits<typename image_view_type::pixel_type>::has_alpha == false);
+            template <typename T, typename image_view_type, typename pixel_type>
+            static bool interpolate(
+                const image_view_type &img,
+                const dlib::vector<T, 2> &p,
+                pixel_type &result
+                )
+            {
+                const long left = static_cast<long>(std::floor(p.x()));
+                const long top = static_cast<long>(std::floor(p.y()));
+                const long right = left + 1;
+                const long bottom = top + 1;
 
-            const long left   = static_cast<long>(std::floor(p.x()));
-            const long top    = static_cast<long>(std::floor(p.y()));
-            const long right  = left+1;
-            const long bottom = top+1;
 
+                // if the interpolation goes outside img 
+                if (!(left >= 0 && top >= 0 && right < img.nc() && bottom < img.nr()))
+                    return false;
 
-            // if the interpolation goes outside img 
-            if (!(left >= 0 && top >= 0 && right < img.nc() && bottom < img.nr()))
-                return false;
+                const double lr_frac = p.x() - left;
+                const double tb_frac = p.y() - top;
 
-            const double lr_frac = p.x() - left;
-            const double tb_frac = p.y() - top;
-
-            double tl, tr, bl, br;
-
-            tl = img[top][left].red;
-            tr = img[top][right].red;
-            bl = img[bottom][left].red;
-            br = img[bottom][right].red;
-            const double red = (1-tb_frac)*((1-lr_frac)*tl + lr_frac*tr) + 
-                                   tb_frac*((1-lr_frac)*bl + lr_frac*br);
-
-            tl = img[top][left].green;
-            tr = img[top][right].green;
-            bl = img[bottom][left].green;
-            br = img[bottom][right].green;
-            const double green = (1-tb_frac)*((1-lr_frac)*tl + lr_frac*tr) + 
-                                   tb_frac*((1-lr_frac)*bl + lr_frac*br);
-
-            tl = img[top][left].blue;
-            tr = img[top][right].blue;
-            bl = img[bottom][left].blue;
-            br = img[bottom][right].blue;
-            const double blue = (1-tb_frac)*((1-lr_frac)*tl + lr_frac*tr) + 
-                                   tb_frac*((1-lr_frac)*bl + lr_frac*br);
-                            
-            rgb_pixel temp;
-            assign_pixel(temp.red, red);
-            assign_pixel(temp.green, green);
-            assign_pixel(temp.blue, blue);
-            assign_pixel(result, temp);
-            return true;
+                blend_component<0>(img, lr_frac, tb_frac, left, top, right, bottom, get_component<0>(result));
```

Comment:
Rather than adding `blend_component` and `get_component` you should use `pixel_to_vector()` and `vector_to_pixel()`.  Then you won't have to duplicate this function several times, but rather have one function that treats pixels generically, knowing nothing about how many channels they have.

Reply:
Nice, with this code is much shorter. Wasn't aware of those utilities :)